bounties_api/analytics/management/commands/timeline_generator.py

```diff
@@ -0,0 +1,222 @@
+    counter = 0
+    accumulator = 0
+
+    for fulfiller in fulfillers:
+        print(fulfiller)
+        fulfillments = time_frame.filter(fulfiller=fulfiller)
+        accepted_fulfillments = fulfillments.filter(accepted=True, accepted_date__lte=accepted_date).count()
+        print(accepted_fulfillments / fulfillments.count())
+        accumulator += accepted_fulfillments / fulfillments.count()
+        counter += 1
+
+    return accumulator / counter
+
+
+def get_avg_fulfillment_amount(time_frame):
+    completed_bounties = filter(lambda bounty: bounty.bountyStage == COMPLETED_STAGE, time_frame)
+    (total, count) = reduce(lambda prev, current: (prev[0] + current.bounty.usd_price, prev[1] + 1),
+                            completed_bounties,
+                            (0, 0))
+
+    return total / count if count > 0 else 0
+
+
+def get_total_fulfillment_amount(time_frame):
+    completed_bounties = filter(lambda bounty: bounty.bountyStage == COMPLETED_STAGE, time_frame)
+    return reduce(lambda prev, current: prev + current.bounty.fulfillmentAmount, completed_bounties, 0)
+
+
+def get_bounty_draft(time_frame):
+    return reduce(add_on(DRAFT_STAGE), time_frame, 0)
+
+
+def get_bounty_active(time_frame):
+    return reduce(add_on(ACTIVE_STAGE), time_frame, 0)
+
+
+def get_bounty_completed(time_frame):
+    return reduce(add_on(COMPLETED_STAGE), time_frame, 0)
+
+
+def get_bounty_expired(time_frame):
+    return reduce(add_on(EXPIRED_STAGE), time_frame, 0)
+
+
+def get_bounty_dead(time_frame):
+    return reduce(add_on(DEAD_STAGE), time_frame, 0)
+
+
+def build_stages(time_frame):
+    """ Build the total for each bounty stage in a given time frame
+
+    1. In a given time frame, are retrieved all distinct bounties
+    2. Get the last stage for the bounty
+    3. Increment the counter for the given stage
+
+    Each position correspond in the stage array correspond to the stage in `constants.py`:
+        DRAFT_STAGE = 0
+        ACTIVE_STAGE = 1
+        DEAD_STAGE = 2
+        COMPLETED_STAGE = 3
+        EXPIRED_STAGE = 4
+
+    returns [total_drafts, total_active, total_dead, total_completed, total_expired]
+
+    **Each total correspond to the last stages of the bounties in the given time frame**
+    """
+    # TODO: Queries optimization
+    unique_bounties = [b['bounty'] for b in time_frame.values('bounty').distinct()]
+
+    stages = [0, 0, 0, 0, 0]
+
+    for bounty_id in unique_bounties:
+        current_stage = time_frame.filter(bounty=bounty_id).aggregate(Max('bountyStage'))['bountyStage__max']
+        stages[current_stage] = stages[current_stage] + 1
+
+    return stages
+
+
+def generate_timeline(time_frame):
+    date = time_frame[1]
+
+    bounties_state_frame = BountyState.objects.filter(change_date__lte=time_frame[1])
+    bounty_created_frame = Bounty.objects.filter(bounty_created__range=time_frame)
+    fulfillment_accepted_frame = Fulfillment.objects.filter(accepted_date__lte=time_frame[1])
```

Comment:
does this grab the null values or no? This is safe?

Reply:
`bounty_created_frame` was deleted in favor to access bounties through BountyState. And nop, such queries doesn't grab null values